### PR TITLE
Include dependently promoted fields in SSA

### PR DESCRIPTION
**Pull request by SingleAccretion**
Allowing for their participation in the SSA-based optimizations.

This changes two significant IR invariants:

1) It is not longer a correctness issue for morph to not mark a dependently promoted struct DNER / multi-reg - we remove the one hard requirement for this determination from the front-end. The soft requirement of CQ and making ref counting happy stays.
2) Uses of promoted locals can now implicitly refer to SSA variables. They are not renamed. The canonical example of such promoted locals are those created for multi-reg returns.

This change also makes `lvIsMultiRegRet` mostly unnecessary, though I plan to scale back its usage in a separate change.

[2nd Round Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=65230&view=ms.vss-build-web.run-extensions-tab)

### CODE REVIEW

**Comment by SingleAccretion** [comment]
@dotnet/jit-contrib 

This will need to be stressed and fuzzed.

**Comment by SingleAccretion** [comment]
The CI results are a bit hard to interpret to known-ness right now (e. g. libraries stress has a little under 900 failures), Fuzzlyn looked clean-ish at least.

**Comment by BruceForstall** [comment]
Stress should be cleaner now.

**Comment by SingleAccretion** [comment]
Some stress (and Fuzzlyn) failures with an assert:
```
    JIT\Directed\Convert\value_numbering_checked_casts_of_constants\value_numbering_checked_casts_of_constants.cmd [FAIL]
      
      Assert failure(PID 4612 [0x00001204], Thread: 3588 [0x0e04]): Assertion failed 'isCandidateVar(fieldVarDsc) == isMultiReg' in 'ValueNumberingCheckedCastsOfConstants:<TestCastingSingleToInt64>g__ConfirmSingleOneFullDecrementUnderByteMaxValueCastToInt64IsFoldedCorrectly|10_27()' during 'Linear scan register alloc' (IL size 92; hash 0xfaebe139; Tier0-MinOpts)
      
          File: D:\a\_work\1\s\src\coreclr\jit\lsrabuild.cpp Line: 1269
          Image: C:\h\w\BB5E09CF\p\corerun.exe
```

**Comment by SingleAccretion** [comment]
Apart from the above, failures looked known.

**Comment by AndyAyersMS** [comment]
Hmm, odd that it got closed -- not sure how this happened.

**Comment by AndyAyersMS** [comment]
Going to bounce this so it runs with ssa checking.

File: src/coreclr/jit/valuenum.cpp [hunk 1]
```diff
@@ -4917,8 +4917,10 @@ void Compiler::fgValueNumberLocalStore(GenTree*             storeNode,
                     // Avoid redundant bitcasts for the common case of a full definition.
                     fieldStoreType = fieldVarDsc->TypeGet();
                 }
+
+                ssize_t      fieldValueOffset = max(0, fieldVarDsc->lvFldOffset - offset);
```
Comment (AndyAyersMS) [inline src/coreclr/jit/valuenum.cpp:4921]: Not clear to me what this `max` is doing.

Comment (SingleAccretion) [reply src/coreclr/jit/valuenum.cpp:4921]: The `max` chooses between the two possible cases:
```
       <---->      - "fieldValueOffset = fieldOffset - offset"
             [--]  - fieldStoreValue, also a full field local
 field offset:
       [---------]  - value
 offset:
[-----------------] - full local
```
```
                     - "fieldValueOffset = 0"
              [-|--] - fieldStoreValue, part of the field local
  field offset: 
                [-----]  - value
          offset:
[----------------------] - full local
```

Comment (AndyAyersMS) [reply src/coreclr/jit/valuenum.cpp:4921]: Perhaps a comment is warranted? 

I guess what confused me was that different aspects of non-entire writes are handled in different places, so it wasn't clear what part of the computation had to happen here.

**Comment by AndyAyersMS** [comment]
libraries jitstress has some known failures; coreclr jitstress was running clean recently.

File: src/coreclr/jit/valuenum.cpp (hunk 1 above)
Comment (SingleAccretion) [reply src/coreclr/jit/valuenum.cpp:4921]: > Perhaps a comment is warranted?

Would this diff make it clearer?
```diff
-                ssize_t      fieldValueOffset = max(0, fieldVarDsc->lvFldOffset - offset);
+                // Calculate offset of this field's value, relative to the entire one.
+                ssize_t      fieldOffset      = fieldVarDsc->lvFldOffset;
+                ssize_t      fieldValueOffset = (fieldOffset < offset) ? 0 : (fieldOffset - offset);
```

> I guess what confused me was that different aspects of non-entire writes are handled in different places, so it wasn't clear what part of the computation had to happen here.

There are two parts to the computations: the "load" part, where we decide what value are we to store, it has as "the base" for the offsets "value", and then there is the "store" part (i. e. `processDef`), where "the base" is the field local. One can notice sizes of both the "load" and "store" parts are the same, so only the offsets need to be calculated differently.

**Comment by SingleAccretion** [comment]
Starting to collect failures.

- Stress
  - PAL `memcpy` tests: not related; opened #77914.
  - `GC/Regressions/Github/Runtime_76219`: does look related, I remember seeing it fail the same way some time ago, and we have an identical, relatively recent, [failure](https://dev.azure.com/dnceng-public/public/_build/results?buildId=65076&view=ms.vss-test-web.build-test-results-tab&runId=1326416&resultId=110857&paneView=dotnet-dnceng.dnceng-anon-build-release-tasks.helix-anon-test-information-tab).
- Fuzzlyn:
  - #77773.

File: src/coreclr/jit/valuenum.cpp (hunk 1 above)
Comment (AndyAyersMS) [reply src/coreclr/jit/valuenum.cpp:4921]: I think it helps, yeah. The other idea I had was to add this as yet another out arg from `gtStoreDefinesField` as it complements `pFieldStoreOffset` but I suppose there are consumers that don't' really care to know this information.

**Comment by SingleAccretion** [comment]
The failures have been accounted for above, I believe all are known.

**Review by AndyAyersMS — APPROVED** [review]

**Comment by kunalspathak** [comment]
> Some (aka a lot of) stress (and Fuzzlyn) failures with an assert:
> 
> ```
>     JIT\Directed\Convert\value_numbering_checked_casts_of_constants\value_numbering_checked_casts_of_constants.cmd [FAIL]
>       
>       Assert failure(PID 4612 [0x00001204], Thread: 3588 [0x0e04]): Assertion failed 'isCandidateVar(fieldVarDsc) == isMultiReg' in 'ValueNumberingCheckedCastsOfConstants:<TestCastingSingleToInt64>g__ConfirmSingleOneFullDecrementUnderByteMaxValueCastToInt64IsFoldedCorrectly|10_27()' during 'Linear scan register alloc' (IL size 92; hash 0xfaebe139; Tier0-MinOpts)
>       
>           File: D:\a\_work\1\s\src\coreclr\jit\lsrabuild.cpp Line: 1269
>           Image: C:\h\w\BB5E09CF\p\corerun.exe
> ```
> 
> Edit: this is fallout from #76263. Fix incoming...

This was also hit in Antigen in yesterday's run. Was this ever fixed?

**Comment by SingleAccretion** [comment]
> Was this ever fixed?

Yes, it was fixed in #77502. Evidently not fully.

**Comment by EgorBo** [comment]
Improvements on arm64-ubuntu: https://github.com/dotnet/perf-autofiling-issues/issues/9679

**Pull request merged**